agent_build_refactored/tools/runner.py

```diff
@@ -0,0 +1,910 @@
+# Copyright 2014-2022 Scalyr Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import argparse
+import dataclasses
+import functools
+import hashlib
+import json
+import os
+import pathlib as pl
+import shutil
+import logging
+import inspect
+import sys
+from typing import Union, Optional, List, Dict, Type
+
+
+from agent_build_refactored.tools.constants import SOURCE_ROOT, DockerPlatformInfo
+from agent_build_refactored.tools import (
+    check_call_with_log,
+    DockerContainer,
+    UniqueDict,
+    check_output_with_log,
+    IN_DOCKER,
+    IN_CICD,
+)
+
+log = logging.getLogger(__name__)
+
+
+def remove_directory_in_docker(path: pl.Path):
+    """
+    Since we produce some artifacts inside docker containers, we may face difficulties with
+    deleting the old ones because they may be created inside the container with the root user.
+    The workaround for that to delegate that deletion to a docker container as well.
+    """
+
+    # In order to be able to remove the whole directory, we mount parent directory.
+    with DockerContainer(
+        name="agent_build_step_trash_remover",
+        image_name="ubuntu:22.04",
+        mounts=[f"{path.parent}:/parent"],
+        command=["rm", "-r", f"/parent/{path.name}"],
+        detached=False,
+    ):
+        pass
+
+
+@dataclasses.dataclass
+class DockerImageSpec:
+    """Simple data class which represents combination of the image name and docker platform."""
+
+    name: str
+    platform: DockerPlatformInfo
+
+    def save_docker_image(self, output_path: pl.Path):
+        """
+        Serialize docker image into file by using 'docker save' command.
+        :param output_path: Result output file.
+        """
+        with output_path.open("wb") as f:
+            check_call_with_log(["docker", "save", self.name], stdout=f)
```

Comment:
I wonder how long this step takes since some Docker images can be quite large and saving it through stdout is probably not the fastest :)

Probably not a big deal, but in case it takes a while, it may be worthwhile looking into various stdout buffer settings to see if we can speed it up by increasing the buffer size or similar.